type_description_interfaces/msg/FieldType.msg

```diff
@@ -0,0 +1,149 @@
+# Represents the type of a field and related meta-data.
+
+# A constant for each type supported according to:
+#   http://design.ros2.org/articles/legacy_interface_definition.html
+# and:
+#   http://design.ros2.org/articles/idl_interface_definition.html
+# Order is loosely coupled to the order of appearance in the IDL 4.2 spec:
+#  https://www.omg.org/spec/IDL/4.2
+
+# Layout of constants across the 0-255 decimal values in the uint8:
+#
+# - 000-031: Primitive types, strings, and reserved space for future primitive types
+# - 032-063: Fixed sized array of primitive and string types
+# - 064-095: Bounded Sequences of primitive and string types
+# - 096-127: Unbounded Sequences of primitive and string types
+# - 128-255: Reserved space for future array/sequence-like types
+
+# Nested type defined in other .msg/.idl files.
+uint8 FIELD_TYPE_NESTED_TYPE = 0
+
+# Basic Types
+# Integer Types
+uint8 FIELD_TYPE_INT8 = 1
+uint8 FIELD_TYPE_UINT8 = 2
+uint8 FIELD_TYPE_INT16 = 3
+uint8 FIELD_TYPE_UINT16 = 4
+uint8 FIELD_TYPE_INT32 = 5
+uint8 FIELD_TYPE_UINT32 = 6
+uint8 FIELD_TYPE_INT64 = 7
+uint8 FIELD_TYPE_UINT64 = 8
+
+# Floating-Point Types
+uint8 FIELD_TYPE_FLOAT = 9
+uint8 FIELD_TYPE_DOUBLE = 10
+uint8 FIELD_TYPE_LONG_DOUBLE = 11
+
+# Char and WChar Types
+uint8 FIELD_TYPE_CHAR = 12
+uint8 FIELD_TYPE_WCHAR = 13
+
+# Boolean Type
+uint8 FIELD_TYPE_BOOLEAN = 14
+
+# Byte/Octet Type
+uint8 FIELD_TYPE_BYTE = 15
+
+# String Types
+uint8 FIELD_TYPE_STRING = 16
+uint8 FIELD_TYPE_WSTRING = 17
+
+# Fixed String Types
+uint8 FIELD_TYPE_FIXED_STRING = 18
+uint8 FIELD_TYPE_FIXED_WSTRING = 19
+
+# Bounded String Types
+uint8 FIELD_TYPE_BOUNDED_STRING = 20
+uint8 FIELD_TYPE_BOUNDED_WSTRING = 21
+
+# Fixed Sized Array Types
+uint8 FIELD_TYPE_NESTED_TYPE_ARRAY = 32
+uint8 FIELD_TYPE_INT8_ARRAY = 33
+uint8 FIELD_TYPE_UINT8_ARRAY = 34
+uint8 FIELD_TYPE_INT16_ARRAY = 35
+uint8 FIELD_TYPE_UINT16_ARRAY = 36
+uint8 FIELD_TYPE_INT32_ARRAY = 37
+uint8 FIELD_TYPE_UINT32_ARRAY = 38
+uint8 FIELD_TYPE_INT64_ARRAY = 39
+uint8 FIELD_TYPE_UINT64_ARRAY = 40
+uint8 FIELD_TYPE_FLOAT_ARRAY = 41
+uint8 FIELD_TYPE_DOUBLE_ARRAY= 42
+uint8 FIELD_TYPE_LONG_DOUBLE_ARRAY= 43
+uint8 FIELD_TYPE_CHAR_ARRAY= 44
+uint8 FIELD_TYPE_WCHAR_ARRAY= 45
+uint8 FIELD_TYPE_BOOLEAN_ARRAY= 46
+uint8 FIELD_TYPE_BYTE_ARRAY= 47
+uint8 FIELD_TYPE_STRING_ARRAY= 48
+uint8 FIELD_TYPE_WSTRING_ARRAY= 49
+uint8 FIELD_TYPE_FIXED_STRING_ARRAY = 50
+uint8 FIELD_TYPE_FIXED_WSTRING_ARRAY = 51
+uint8 FIELD_TYPE_BOUNDED_STRING_ARRAY = 52
+uint8 FIELD_TYPE_BOUNDED_WSTRING_ARRAY = 53
+
+# Bounded Sequence Types
+uint8 FIELD_TYPE_NESTED_TYPE_BOUNDED_SEQUENCE = 64
```

Comment:
In translating these values outside this code, I implemented as "first 21, then add a constant based on type". Maybe this would be easier as two fields? The actual payload would be larger but it would reduce constant definitions by 4x.

```
uint8 value_type   # the first 21 values defined here
uint8 sequence_type   # value, array, bounded_sequence, or unbounded_sequence 
```

Reply:
I was optimizing for size. Also, there's not perfect overlap I think. I'd have to double check that. (cases where there are values that cannot be sequences as well).

Reply:
I'm not sure optimizing for fewer constants is that important, but I could be wrong.